File: Ruby/signature.rb
require 'base64'
require 'openssl'

# In rack apps that include Rails and Sinatra,
# raw_data is request.raw_body which only can be read once.

# Example:
# error!("Authorization is invalid", 401) unless MeiqiaToolsSigner.get(request.raw_body, secrect_key) == headers["Authorization"]
class MeiqiaSignTools
  def initialize(key)
    @key = key
  end

  def sign(raw_body)
    hash_str = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha1'), @key, raw_body)
    sign_str = Base64.urlsafe_encode64(hash_str)
    "meiqia_sign:#{sign_str}"
  end
end

File: PHP/signature.php
<?php
// PHP 5 >= 5.1.2
// raw_data is http_get_request_body() that get request body as string
// Example:
// $body = http_get_request_body()  「如果 http_get_request_body() 不能使用，可以用 $body = @file_get_contents('php://input')」
// $dt = new DTSigner($secrect_key)
// $headers = apache_request_headers()
// if($dt->sign($body) != $headers['Authorization']){
//
//}

  class DTSigner {

    private $raw_data;
    private $key;

    public function __construct($key) {
        $this->key = $key;
    }

    public function sign($raw_data) {
        $hash_str = hash_hmac('sha1', $raw_data, $this->key);
        $sign_str = strtr(base64_encode($hash_str), '+/', '-_');
        return 'meiqia_sign:'.$sign_str;
    }

  }
?>
